packages/coding-standards/Jetpack/Sniffs/InlineDocs/HooksMustHaveDocblockSniff.php

```diff
@@ -0,0 +1,157 @@
+	 *  )
+	 * )
+	 *
+	 * @return array
+	 */
+	public function getGroups() {
+		return array(
+			'hooks' => array(
+				'functions' => array_keys( $this->hook_functions ),
+			),
+		);
+	}
+
+	/**
+	 * Process a matched token.
+	 *
+	 * @since 1.0.0 Logic split off from the `process_token()` method.
+	 *
+	 * @param int    $stack_ptr       The position of the current token in the stack.
+	 * @param string $group_name      The name of the group which was matched.
+	 * @param string $matched_content The token content (function name) which was matched.
+	 *
+	 * @return int|void Integer stack pointer to skip forward or void to continue
+	 *                  normal file processing.
+	 */
+	public function process_matched_token( $stack_ptr, $group_name, $matched_content ) {
+
+		$func_open_paren_token = $this->phpcsFile->findNext( Tokens::$emptyTokens, ( $stack_ptr + 1 ), null, true );
+		if ( false === $func_open_paren_token
+			|| \T_OPEN_PARENTHESIS !== $this->tokens[ $func_open_paren_token ]['code']
+			|| ! isset( $this->tokens[ $func_open_paren_token ]['parenthesis_closer'] )
+		) {
+			// Live coding, parse error or not a function call.
+			return;
+		}
+
+		$previous_comment = $this->phpcsFile->findPrevious( Tokens::$commentTokens, ( $stack_ptr - 1 ) );
+
+		if ( false !== $previous_comment ) {
+			/*
+			 * Check to determine if there is a comment immediately preceding the function call.
+			 */
+			if ( ( $this->tokens[ $previous_comment ]['line'] + 1 ) !== $this->tokens[ $stack_ptr ]['line'] ) {
+				$this->phpcsFile->addError(
+					'The inline documentation for a hook must be on the line immediately before the function call.',
+					$stack_ptr,
+					'DocMustBePreceding'
+				);
+			}
+
+			/*
+			 * Check that the comment starts is a docblock.
+			 */
+			if ( \T_DOC_COMMENT_CLOSE_TAG !== $this->tokens[ $previous_comment ]['code'] ) {
+				$this->phpcsFile->addError(
+					'Hooks must include a docblock with /** formatting */',
+					$stack_ptr,
+					'NoDocblockFound'
+				);
+				// return; Do we need to return here? Can we keep going?
+			}
+
+			/*
+			 * Process docblock tags.
+			 */
+			$comment_end   = $previous_comment;
+			$comment_start = ( isset( $this->tokens[ $comment_end ]['comment_opener'] ) ) ? $this->tokens[ $comment_end ]['comment_opener'] : false;
+			$has           = array(
+				'module' => false,
+				'since'  => false,
+			);
+
+			// The comment isn't a docblock, so we're going to stop here.
+			if ( ! $comment_start ) {
+				return;
+			}
+
+			$string = $this->phpcsFile->findNext( T_DOC_COMMENT_STRING, $comment_start, $comment_end );
+			// If the call is documented elsewhere, stop here.
+			if ( 0 === strpos( $this->tokens[ $string ]['content'], 'This filter is documented in' ) ) {
+				return;
+			}
+
+			foreach ( $this->tokens[ $comment_start ]['comment_tags'] as $tag ) {
+				// Is the next tag of the docblock the "@module" tag?
+				if ( '@module' === $this->tokens[ $tag ]['content'] ) {
```

Comment:
Thinking about how this might get pulled up into WPCS in the future, I wonder if there's value in splitting this into two sniffs - one that checks whether there's a DocBlock (and even some WP-typical tags in it) for hooks, and another sniff which looks for the Jetpack-specific expected tags.

Reply:
Makes sense. First pass at splitting them with the idea that, if this lands in WPCS, other standards can extend it for specific requirements like Jetpack. ede33130c